source/opt/aggressive_dead_code_elim_pass.cpp

```diff
@@ -123,40 +129,193 @@ void AggressiveDCEPass::ProcessLoad(uint32_t varId) {
+}
+
+bool AggressiveDCEPass::IsStructuredIfHeader(ir::BasicBlock* bp,
+      ir::Instruction** mergeInst, ir::Instruction** branchInst,
+      uint32_t* mergeBlockId) {
+  auto ii = bp->end();
+  --ii;
+  if (ii->opcode() != SpvOpBranchConditional)
+    return false;
+  if (ii == bp->begin())
+    return false;
+  *branchInst = &*ii;
+  --ii;
+  if (ii->opcode() != SpvOpSelectionMerge)
+    return false;
+  *mergeInst = &*ii;
+  *mergeBlockId = ii->GetSingleWordInOperand(kSelectionMergeMergeBlockIdInIdx);
+  return true;
+}
+
+void AggressiveDCEPass::ComputeBlock2BranchMaps(
+      std::list<ir::BasicBlock*>& structuredOrder) {
+  block2merge_.clear();
+  block2branch_.clear();
+  std::stack<ir::Instruction*> currentMergeInst;
+  std::stack<ir::Instruction*> currentBranchInst;
+  std::stack<uint32_t> currentMergeBlockId;
+  currentMergeInst.push(nullptr);
+  currentBranchInst.push(nullptr);
+  currentMergeBlockId.push(0);
+  for (auto bi = structuredOrder.begin(); bi != structuredOrder.end(); ++bi) {
+    if ((*bi)->id() == currentMergeBlockId.top()) {
+      currentMergeBlockId.pop();
+      currentMergeInst.pop();
+      currentBranchInst.pop();
+    }
+    block2merge_[*bi] = currentMergeInst.top();
+    block2branch_[*bi] = currentBranchInst.top();
+    ir::Instruction* mergeInst;
+    ir::Instruction* branchInst;
+    uint32_t mergeBlockId;
+    if (IsStructuredIfHeader(*bi, &mergeInst, &branchInst, &mergeBlockId)) {
+      currentMergeBlockId.push(mergeBlockId);
+      currentMergeInst.push(mergeInst);
+      currentBranchInst.push(branchInst);
+    }
+  }
+}
+
+void AggressiveDCEPass::ComputeInst2BlockMap(ir::Function* func) {
+  for (auto& blk : *func) {
+    blk.ForEachInst([&blk,this](ir::Instruction* ip) {
+      inst2block_[ip] = &blk;
+    });
+  }
+}
+
+void AggressiveDCEPass::AddBranch(uint32_t labelId, ir::BasicBlock* bp) {
+  std::unique_ptr<ir::Instruction> newBranch(
+    new ir::Instruction(SpvOpBranch, 0, 0,
+        {{spv_operand_type_t::SPV_OPERAND_TYPE_ID, {labelId}}}));
+  def_use_mgr_->AnalyzeInstDefUse(&*newBranch);
+  bp->AddInstruction(std::move(newBranch));
+}
+
 bool AggressiveDCEPass::AggressiveDCE(ir::Function* func) {
+  // Compute map from instruction to block
+  ComputeInst2BlockMap(func);
+  // Compute map from block to controlling conditional branch
+  std::list<ir::BasicBlock*> structuredOrder;
+  ComputeStructuredOrder(func, &structuredOrder);
+  ComputeBlock2BranchMaps(structuredOrder);
   bool modified = false;
   // Add all control flow and instructions with external side effects 
   // to worklist
   // TODO(greg-lunarg): Handle Frexp, Modf more optimally
   call_in_func_ = false;
   func_is_entry_point_ = false;
   private_stores_.clear();
-  for (auto& blk : *func) {
-    for (auto& inst : blk) {
-      uint32_t op = inst.opcode();
+  std::stack<bool> skip_control_flow;
+  std::stack<uint32_t> currentMergeBlockId;
+  skip_control_flow.push(false);
+  currentMergeBlockId.push(false);
```

Comment:
Did you mean to push 0, signifiying no merge block at all?  The false is misleading.

Reply:
Yes. Surprised I did that.